File: leetcode.com/python/212_Word_Search_II.py
# Getting "Maximum recursion depth exceeded. WHy the fuck?  Did't wunderstand why
from collections import defaultdict
class TrieNode:
    def __init__(self):
        self.children = collections.defaultdict(TrieNode)
        self.isWord = False
        self.fullWord = None

class Trie:
    def __init__(self):
        self.root = TrieNode()

    def addWords(self, words):
        for word in words:
            currentNode = self.root
            for char in word:
                if not currentNode.children[char]:
                    currentNode.children[char] = TrieNode()
                currentNode = currentNode.children[char]
            currentNode.isWord = True
            currentNode.fullWord = word

    def searchWord(self, word):
        cirrentWord = self.root
        for char in word:
            if not cirrentWord.children[char]:
                return False
            cirrentWord = cirrentWord.children[char]
        return True

class Solution(object):
    def findWords(self, board, words):
        """
        :type board: List[List[str]]
        :type words: List[str]
        :rtype: List[str]
        """
        allWord = []
        trie = Trie()
        trie.addWords(words)
        for row in range(len(board)):
            for col in range(len(board[0])):
                self.searchBoardDFS(board, row, col, trie.root, allWord)
        return allWord

    def searchBoardDFS(self, board, row, col, currentNode, allWord):
        if currentNode.isWord:
            allWord.append(currentNode.fullWord)
            currentNode.isWord = False # to prevent from adding duplicate word

        if row < 0 or row >= len(board) or col < 0 or col >= len(board[0]):
            return

        currentChar = board[row][col]
        currentNode = currentNode.children[currentChar]

        if not currentNode: # currentChar doesn't exist
            return

        board[row][col] = "#" # to prevent going into loop and revisiting
        self.searchBoardDFS(board, row+1, col, currentNode, allWord)
        self.searchBoardDFS(board, row-1, col, currentNode, allWord)
        self.searchBoardDFS(board, row, col+1, currentNode, allWord)
        self.searchBoardDFS(board, row, col-1, currentNode, allWord)
        board[row][col] = currentChar  # Backtrack




# https://tinyurl.com/y7td7eag
class TrieNode():
    def __init__(self):
        self.children = collections.defaultdict(TrieNode)
        self.isWord = False


class Trie():
    def __init__(self):
        self.root = TrieNode()

    def insert(self, word):
        node = self.root
        for w in word:
            node = node.children[w]
        node.isWord = True

    def search(self, word):
        node = self.root
        for w in word:
            node = node.children.get(w)
            if not node:
                return False
        return node.isWord


class Solution(object):
    def findWords(self, board, words):
        res = []
        trie = Trie()
        node = trie.root
        for w in words:
            trie.insert(w)
        for i in range(len(board)):
            for j in range(len(board[0])):
                self.dfs(board, node, i, j, "", res)
        return res

    def dfs(self, board, node, i, j, path, res):
        if node.isWord:
            res.append(path)
            node.isWord = False
        if i < 0 or i >= len(board) or j < 0 or j >= len(board[0]):
            return
        tmp = board[i][j]
        node = node.children.get(tmp)
        if not node:
            return
        board[i][j] = "#"
        self.dfs(board, node, i + 1, j, path + tmp, res)
        self.dfs(board, node, i - 1, j, path + tmp, res)
        self.dfs(board, node, i, j - 1, path + tmp, res)
        self.dfs(board, node, i, j + 1, path + tmp, res)
        board[i][j] = tmp

File: leetcode.com/swift/DS_ALGO_PRAC.playground/Sources/swift/212_Word_Search_II.swift

import Foundation

class TrieNode {
    var children = [Character: TrieNode]()
    var isWord = false
    var word: String?
}

class Trie {
    var root = TrieNode()
    
    func insert(_ word: String) {
        var node = self.root
        for char in word {
            if node.children[char] == nil {
                node.children[char] = TrieNode()
                
            }
            node = node.children[char]!
        }
        node.isWord = true
        node.word = word
    }
}

class Solution {
    func findWords(_ board: [[Character]], _ words: [String]) -> [String] {
        var presentWords = [String]()
        var trie = Trie()
        for word in words {
            trie.insert(word)
        }

        var copyBoard = board
        for i in 0..<board.count {
            for j in 0..<board[0].count {
                dfsWordSearch(&copyBoard, &trie.root, i, j, &presentWords)
            }
        }
        return presentWords
    }
    
    func dfsWordSearch(_ board: inout [[Character]], _ currentNode: inout TrieNode, _ i: Int, _ j: Int, _ foundWords: inout [String]) {
        if currentNode.isWord {
            foundWords.append(currentNode.word!)
            currentNode.isWord = false
        }
        if i < 0 || i >= board.count || j < 0 || j >= board[0].count || board[i][j] == "#" {
            return
        }
        let currChar = board[i][j]
        guard var nextNode = currentNode.children[currChar] else {
            return
        }
        board[i][j] = Character("#")
        for (nexti, nextj) in [(i-1, j), (i+1, j), (i, j-1), (i, j+1)] {
            dfsWordSearch(&board, &nextNode, nexti, nextj, &foundWords)
        }
        board[i][j] = currChar
    }
}
